opentelemetry-sdk/src/opentelemetry/sdk/_metrics/instrument.py

```diff
@@ -30,131 +29,107 @@
+
+
+class _Synchronous(_Instrument):
     def __init__(
         self,
         instrumentation_info: InstrumentationInfo,
         name: str,
-        *args,
         unit: str = "",
         description: str = "",
     ):
         self._instrumentation_info = instrumentation_info
-        super().__init__(name, *args, unit=unit, description=description)
+        super().__init__(name, unit=unit, description=description)
 
 
-class Counter(_Instrument, APICounter):
+class _Asynchronous(_Instrument):
     def __init__(
         self,
         instrumentation_info: InstrumentationInfo,
         name: str,
+        callback: CallbackT,
         unit: str = "",
         description: str = "",
     ):
-        super().__init__(
-            instrumentation_info,
-            name,
-            unit=unit,
-            description=description,
-        )
-
-    def add(self, amount, attributes=None):
-        # FIXME check that the amount is non negative
-        pass
 
+        self._instrumentation_info = instrumentation_info
+        super().__init__(name, callback, unit=unit, description=description)
 
-class UpDownCounter(_Instrument, APIUpDownCounter):
-    def __init__(
-        self,
-        instrumentation_info: InstrumentationInfo,
-        name: str,
-        unit: str = "",
-        description: str = "",
-    ):
-        super().__init__(
-            instrumentation_info,
-            name,
-            unit=unit,
-            description=description,
-        )
-
-    def add(self, amount, attributes=None):
-        pass
+        self._callback = callback
 
+        if isinstance(callback, Generator):
 
-class ObservableCounter(_Instrument, APIObservableCounter):
-    def __init__(
-        self,
-        instrumentation_info: InstrumentationInfo,
-        name: str,
-        callback: TCallback,
-        unit: str = "",
-        description: str = "",
-    ):
-        super().__init__(
-            instrumentation_info,
-            name,
-            callback,
-            unit=unit,
-            description=description,
-        )
+            def inner() -> Iterable[Measurement]:
+                return next(callback)
 
+            self._callback = inner
 
-class ObservableUpDownCounter(_Instrument, APIObservableUpDownCounter):
-    def __init__(
-        self,
-        instrumentation_info: InstrumentationInfo,
-        name: str,
-        callback: TCallback,
-        unit: str = "",
-        description: str = "",
+    @property
+    def callback(self) -> CallbackT:
+        return self._callback
+
+
+class Counter(_Synchronous, APICounter):
+    @property
+    def default_aggregation(self):
+        return SumAggregation
+
+    def add(
+        self, amount: Union[int, float], attributes: Dict[str, str] = None
     ):
-        super().__init__(
-            instrumentation_info,
-            name,
-            callback,
-            unit=unit,
-            description=description,
-        )
+        if amount < 0:
```

Comment:
Curious as to why this behavior is defined in the sdk and not the api? Shouldn't all implementations of `counter.add` not take in negative increments?

Reply:
We received some [feedback](https://github.com/open-telemetry/opentelemetry-python/pull/1887#discussion_r711360943) time ago regarding where do we put the checks.

In my opinion, the checks should be in the API as you suggest, for the exact reason you mention. I think the discussion linked above was also influenced by the fact that we were raising an exception in the code as it was before and that is related to [this](https://github.com/open-telemetry/opentelemetry-python/pull/2164#issuecomment-932432097) discussion we had.

I would suggest that we move these checks to the API, and if the check fails, an exception be raised. That exception would be handled by an [error handling mechanism](https://github.com/open-telemetry/opentelemetry-python/pull/2244) that we need to add in the future.

Reply:
@lzchen in order to move this PR forward, I am marking this conversation as resolved. I have opened an issue to track this concern (which fortunately can be addressed in a later PR). Sorry for proceeding like this, I would have preferred to discuss this with you before but you are currently out of office :slightly_smiling_face: